dev/core#663 - Use InnoDB engine for extended log tables

pfigel opened the pull request:
Overview
----------------------------------------
CiviCRM's optional detailed logging feature creates a log table to mirror every table in the `civicrm_%` namespace. To preserve disk space, log tables use the `ARCHIVE` engine by default. `ARCHIVE` has a number of limitations, most importantly:
- No index support
- No support for `UPDATE` or `DELETE`
- No crash-safety or transaction support

This has lead to a number of performance, reliability and durability issues in the past. It is possible to use different storage engines by installing extensions, but CiviCRM should offer sane and safe defaults.

Before
----------------------------------------
Detailed log tables use the `ARCHIVE` engine if available.

After
----------------------------------------
Detailed log tables on new installations use the `InnoDB` engine by default. Log tables created at some point after this version was installed also use the `InnoDB` engine.

Existing log tables will continue to use the `ARCHIVE` table.

Site administrators may convert existing log tables using the `System.updatelogtables` API with `forceEngineMigration` set.

Technical Details
----------------------------------------
Some tests were added to cover the new behaviour. Migration code will not trigger if an engine was set by the hook, so `ARCHIVE` could still be used that way.

Comments
----------------------------------------
An update for the sysadmin guide was prepared in https://github.com/civicrm/civicrm-sysadmin-guide/pull/167, including notes on migration and a way to restore the old default (should anyone want to do that). It assumes we're targeting 5.16 for this.

xurizaemon commented on CRM/Logging/Schema.php:
```diff
@@ -302,7 +302,7 @@ public function fixSchemaDifferences($enableLogging = FALSE) {
+   * to INNODB if engine has not been deliberately set (by hook) and not dropping
    * indexes. Sysadmin will need to manually intervene to revert to defaults.
    */
   public function updateLogTableSchema() {
```
Can I suggest that here and in the naming of the API call `System.updatelogtables`, we are explicit about what this actually does? Eg `convertDataLoggingTablesFromArchiveToInnoDB()` is explicit, where "updateLogTables" could mean many things.

* In future there will be other updates, and a function named like this will occupy a namingspace that also fits other functions.
* Who wouldn't run a function called "update"? Sounds reasonable, but [as @Stoob points out](https://lab.civicrm.org/dev/core/issues/663#note_17556), this process has an impact on a functioning site.
* CiviCRM has a pattern of recycling terminology in multiple spaces (I count [six different meanings for "log"](https://gist.github.com/xurizaemon/01a53b0e63b73f88de61dca5d9afe80a)), I think it's helpful for both DX and UX if we refer to things using clear terminology.

xurizaemon replied:
Comment above was written thinking you'd added the function in this PR - on second view this need not block your proposed changes (but please consider it where adding new methods etc).

pfigel replied:
I agree that the current naming is confusing - I've spent time in this code in the past, but I still get `System.updatelogtables` mixed up with `System.createmissinglogtables` all the time.

What makes it a bit tricky is that, in addition to changing the engine schema, the function/API will potentially also do the following things:

- Change the column definition of `log_conn_id`
- Create any missing indexes (which may have been added via `hook_civicrm_alterLogTables`)

I'm okay with renaming the existing API if we can come up with a less confusing name for the API that encapsulates all of that.

eileenmcnaughton commented:
@pfigel I feel like we would normally chunk things like this over a couple of releases with one release just changing the default behaviour for new installs / new enable loggings & the next chunk addressing existing sites

pfigel commented:
@eileenmcnaughton I've extracted the system check/upgrade prompt for a future PR and changed `System.updatelogtables` to only trigger migration if a new `forceEngineMigration` parameter is set.

This is now ready for review. I've opened a corresponding doc update in https://github.com/civicrm/civicrm-sysadmin-guide/pull/167.

pfigel commented:
test this please

eileenmcnaughton commented on CRM/Logging/Schema.php:
```diff
@@ -302,23 +310,36 @@ public function fixSchemaDifferences($enableLogging = FALSE) {
    */
   public function updateLogTableSchema($params = []) {
     isset($params['updateChangedEngineConfig']) ? NULL : $params['updateChangedEngineConfig'] = FALSE;
+    isset($params['forceEngineMigration']) ? NULL : $params['forceEngineMigration'] = FALSE;
```
Stylistically I have a problem with this - 

```
$params['forceEngineMigration'] =  isset($params['forceEngineMigration']) ? $params['forceEngineMigration'] : FALSE ;

```

would be more consistent with our style elsewhere (I realise the style could be argued but that's consistent with our general code style). However there are 2 other options

1) null coalescing operator now we are on 7.x....

```
$params['formEngineMigration'] = $params['formEngineMigration'] ?? FALSE;
```

I would also consider setting as an api.default value

eileenmcnaughton commented:
@Stoob ping

eileenmcnaughton commented:
This is good from a code point of view - I made minor comments - @Stoob is going to try to r-run it & if that is good we can merge

pfigel commented:
@eileenmcnaughton @Stoob I added API defaults for `forceEngineMigration` (and `updateChangedEngineConfig`, to keep it consistent).

mattwire commented:
@pfigel If we're already using https://github.com/eileenmcnaughton/nz.co.fuzion.innodbtriggers and have already converted our log tables to innodb then applying this patch and disabling that extension should be fine?

pfigel commented:
@mattwire generally yes, although this change does not enable InnoDB compression (`ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4`), which `nz.co.fuzion.innodbtriggers` does, so it's not a drop-in replacement.

The behaviour would be like this:
1. For *new* installations (or installations that have never had logging enabled): InnoDB is used
2. For existing installations *without* `nz.co.fuzion.innodbtriggers` or similar: `ARCHIVE` is used for existing log tables, `InnoDB` for any new ones. Migration to `InnoDB` may be forced via `forceEngineMigration`.
3. For existing installations *with* `nz.co.fuzion.innodbtriggers` or similar: The engine/config set by the extension wins.
4. For existing installations that have *previously* used `nz.co.fuzion.innodbtriggers` or similar: Existing log tables will remain as-is, new ones will not use the engine config `ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=4` (i.e. will not be compressed).

New tables are generally any tables in the `civicrm_%` namespace that are added by core or an extensions, or a new custom field set created at some point after this change.

seamuslee001 commented:
I have tested this and confirmed that without this change log tables are created as Archive, with this change they are created with INNODB, i have also confirmed that this change by default doesn't change the log tables themselves without the `forceEngineMigration` parameter and confirmed when that is set the tables are correctly converted to INNODB. This looks good to me I'm going to merge based on my Testing and EIleen's comment

eileenmcnaughton commented:
Yay - Archive is on the way out!

Stoob commented:
My test site successfully implemented nz.co.fuzion.innodbtriggers on about 150k records.  I'm eager to give this patch a try out too.  I can rebuild my dev site if needed @eileenmcnaughton

eileenmcnaughton commented:
@Stoob so I think this uses INNODB by default - but I don't think it adds any indexes by default - which you'd likely want for 150k records - so nz.co.fuzion.innodbtriggers may still make sense for you

Stoob commented:
Ok, thanks for the clarification @eileenmcnaughton.  So the patch makes the table engine conversion, but does not add indexes.  The extension adds indexes.  For large installs, the extension adds value.

eileenmcnaughton commented:
@Stoob yes that is my understanding
